src/main/scala/scalafix/sbt/ScalafixPlugin.scala

```diff
@@ -97,6 +97,25 @@ object ScalafixPlugin extends AutoPlugin {
     def scalafixConfigSettings(config: Configuration): Seq[Def.Setting[_]] =
       inConfig(config)(
         Seq(
+          scalacOptions := {
+            val options = scalacOptions.value
+            if (!scalafixInvokedAlone.value) options
+            else
+              options.filterNot { option =>
+                scalacOptionsToRelax.exists(_.matcher(option).matches)
+              }
```

Comment:
This `scalafixInvokedAlone` can be hard to debug when you don't know it exists. I think it is very surprising that the `scalafix` task does not behave the same when it is invoked alone or not.

Why not scoping the filtered `scalacOptions` to the `scalafix` task?
```
scalafix / scalacOptions := scalacOptions.value.filterNot(...)
```

Reply:
Thanks for the review @adpi2 !

> I think it is very surprising that the scalafix task does not behave the same when it is invoked alone or not.

Indeed. I guess we could just fail fast if we detect other tasks? As described in the notes of https://github.com/scalacenter/sbt-scalafix/pull/196/commits/abb7e1bb3504f3ff81e1bb71c0331293c2540812, `scalafix[All]` being an input task, it is a very unlikely scenario anyway. I just added cdffae54a5e4585c5a5770b5402e38fc4fc85f62.

> Why not scoping the filtered scalacOptions to the scalafix task?

The problem is that we want to affect the regular Zinc compilation which does not look up `scalafix / scalacOptions`.

1. When semantic rules are requested, `scalafix` needs classes and semanticdb information to run.
1. It uses the [regular compilation output directory](https://github.com/scalacenter/sbt-scalafix/blob/1eeafe2df0c86a3cf4d90137cc3d2c00adb8b407/src/main/scala/scalafix/sbt/ScalafixPlugin.scala#L443) (as opposed to a separate compilation) to leverage the incremental compilations requested by the user in their normal workflow.
1. [An explicit call to `compileIncremental` (via `manipulateBytecode`)](https://github.com/scalacenter/sbt-scalafix/blob/1eeafe2df0c86a3cf4d90137cc3d2c00adb8b407/src/main/scala/scalafix/sbt/ScalafixPlugin.scala#L453-L454) is done, to make sure that the compiler output is up to date, failing before calling Scalafix if the compilation is not successful.
1. To avoid getting warning as errors that would prevent generation of class and semanticdb files and defeat the purpose of Scalafix migration rules, this PR alters the `compileIncremental`, transitively invoked by `scalafix`.

To follow you suggestion while still honoring the goals of (2) & (4), I tried to define a separate, more-relaxed `compileIncremental`:
```scala
// pseudo code showing that relaxed compilation would only be triggered in case of failure of the regular one
scalafix := scalafixInputTaskImpl
  .dependsOn(compileIncremental.recoverWith(scalafix / compileIncremental)).value

inTask(scalafix)(
  Seq(
    scalacOptions -= "-Xfatal-warnings",
    compileIncremental := compileIncrementalTask,
  ) ++ compileInputsSettings // compileInputs looks up compileOptions which looks up scalacOptions
)
```
But the problem is that [`compileInputs` is explicitly looked up within `compile`](https://github.com/sbt/sbt/blob/f155bfd1f3bdcea4b66c0480c1cef79b74410559/main/src/main/scala/sbt/Defaults.scala#L2262), so the enclosing `inTask(scalafix)` has no effect where we would like it to: `scalafix / compileInputs` and thus `scalafix / scalacOptions` are not looked up.

Any idea?

Reply:
Thanks for the explanation, it was not clear to me that the `scalafix` task could trigger a compilation through the uncommon dependency on `manipulateBytecode`.

So you are right, the `scalafix / compileInputs` are not looked up by the regular `compileIncremental` task and I don't see any solution.

The fail-fast strategy makes things a bit better, but I am still worried about the current implementation which is quite convoluted. Also I have been able to reproduce some bug while testing:

1. In a project that contains at least 2 sources, the top one depending on the bottom one.
2. In the top source, add a piece of code that produces a warning
```
def foo {}
```
3. Run compile, it fails
```
sbt:root> compile
[info] compiling 1 Scala source to /home/piquerez/scalacenter/sbt-scalafix-196/target/scala-2.13/classes ...
[error] /home/piquerez/scalacenter/sbt-scalafix-196/src/main/scala/foo/Main.scala:5:11: procedure syntax is deprecated: instead, add `: Unit =` to explicitly declare `foo`'s return type
[error]   def foo {}
[error]           ^
[error] No warnings can be incurred under -Werror.
[error] two errors found
[error] (Compile / compileIncremental) Compilation failed
[error] Total time: 3 s, completed Mar 8, 2021, 11:04:47 AM
```
4. Run scalafixAll, it should compile one source successfully
```
sbt:root> scalafixAll
[info] compiling 1 Scala source to /home/piquerez/scalacenter/sbt-scalafix-196/target/scala-2.13/classes ...
[warn] /home/piquerez/scalacenter/sbt-scalafix-196/src/main/scala/foo/Main.scala:5:11: procedure syntax is deprecated: instead, add `: Unit =` to explicitly declare `foo`'s return type
[warn]   def foo {}
[warn]           ^
[warn] one warning found
[info] Running scalafix on 2 Scala sources
[success] Total time: 1 s, completed Mar 8, 2021, 11:05:41 AM
```
5. Run compile again
```
sbt:root> compile
[info] compiling 2 Scala sources to /home/piquerez/scalacenter/sbt-scalafix-196/target/scala-2.13/classes ...
[error] /home/piquerez/scalacenter/sbt-scalafix-196/src/main/scala/foo/Main.scala:5:11: procedure syntax is deprecated: instead, add `: Unit =` to explicitly declare `foo`'s return type
[error]   def foo {}
[error]           ^
[error] No warnings can be incurred under -Werror.
[error] two errors found
[error] (Compile / compileIncremental) Compilation failed
[error] Total time: 0 s, completed Mar 8, 2021, 11:06:35 AM
```
The 2 sources are being recompiled which means that the Zinc cache has been completely invalidated, for no reason. In a bigger project all files are being recompiled.

Reply:
An alternative solution could be to introduce a `scalafixCompile` task that duplicate the `compile` logic using the same Zinc cache and the same semanticdb target to leverage the incremental compilation.

The benefits would be:
1. Break the cyclic dependency when `scalaOnCompile := true`
2. Make the relax scalacOptions implementation more simple: `compile` depends on `compile / scalacOptions` while `scalafixCompile` depends on `scalafixCompile / scalacOptions`

But also it has some drawbacks:
1. Can be hard to implement: in particular reusing the Zinc cache without altering it
2. Duplication of code: hard to maintain if the sbt implementation evolve

@bjaglin I let you decide if this alternative is worth trying or not. Otherwise I would be okay to validate the current solution.

Reply:
> Also I have been able to reproduce some bug while testing

Good catch. I just added 8f407e1aeb7bf994f290138de5d2e7c19355da79 which should cover this. I couldn't find a simple way to assert this behavior in the scripted, so I made sure the zinc analysis file was not saved for scalafix invocations, which should prevent other related bugs.

Reply:
> An alternative solution could be to introduce a scalafixCompile task that duplicate the compile logic using the same Zinc cache and the same semanticdb target to leverage the incremental compilation.

I initially ruled out that option as I thought it was would be too complicated, and unfortunately, after digging a bit more into it, I am not very hopeful.
1. As you mention in drawback (2), it would be very hard to track sbt changes and support a decent range of sbt versions, just looking at how `compileIncrementalTaskImpl` has evolved in the last months with BSP & pipeling. This is a case for [derived settings](https://github.com/sbt/sbt/discussions/6323) I believe.
1. Even though the current solution can be confusing as it changes the behavior of the Zinc compilation based on the invoked task, checking the logs of `compileIncremental` after a scalafix run does show a bit what's going on. The output of a potential `scalafixCompileIncremental` (classes & semantidcb files) would be shared with the regular `compileIncremental`, making it maybe a bit more obscure and subject to races.
1. Apart from relaxing the scalac invocation in the scope(s) where scalafix was specifically invoked, parent configs/projects (via `dependencyClasspath`) should probably also be relaxed like the current implementation of this PR does. Running a migration script fixing warnings would require many successive runs if we fail to do so, defeating the UX goal of this PR.

Reply:
Yet another approach to this PR would be to be resilient to a failing `compileIncremental` and start scalafix anyway against stale/missing class files, relying on a [on-demand, relaxed compilation in scalafix for semanticdb info](https://github.com/scalacenter/scalafix/blob/fd1ceef8f2a20592e79eae743bef3c02eadccad6/scalafix-cli/src/main/scala/scalafix/internal/v1/MainOps.scala#L338). At the moment the `SymbolTable` used by scalafix does not make use of that on-demand compilation though (and I don't know if we even can?), so this would be very hacky.

Reply:
@adpi2 happy to hear your thoughts in case I missed something